[#151015901] Product page update & onboarding

stephenjoe1 opened the pull request:
# What

This PR is the result of several stories that have gotten a bit mixed up, but the main changes are

* Merges in many of the product page tweaks that have been floating around in the `paas-proto` experiments for some time.
* Migrates the backend from a static site to a dynamic ruby/sinatra app to enable us to do form submission and API calls to things like Deskpro. (This one done by using the previously created paas-submit application as the backend and letting it serve the erb files). 
* Adds the new "Getting Started" onboarding flow
* Adds a form/wizard for users to request a PaaS account
* Adds a generic support/contact-us form for creating tickets

# Why is this not squashed!?

Because the path to squashery is not yet clear.

...and I am expecting changes from review before I make it prettier.

# How to review

* Ensure you can run the app locally: `make dev`
* Ensure you can run the tests `make test` and they pass
* You may find it easier to code review this application in it's entirety at this point rather than per commit. So..
    * Code review of the sinatra backend (app.rb)
    * Check for any of the common web app mistakes (escaping/injection stuff in templating etc)
    * 
* Proof read content, especially on:
    * `/` (home)
    * `/signup`
    * `/contact-us`
* Ensure forms work (you will need to run the app locally with deskpro keys - see readme)
    * ...with js enabled: forms should display as 'wizard' with validation as you go
    * ...with js disabled: forms should display as a single big form with validation on submit
* There is a PR for updating the product-page deployment in release-ci: https://github.com/alphagov/paas-release-ci/pull/38

# Who can review

not @chrisfarms nor @samcrang nor @stephenjoe1

chrisfarms commented on views/signup.erb:
```diff
@@ -0,0 +1,213 @@
+				<div class="step step--about-you step--all">
+					<h2 class="heading-medium">Your contact details</h2>
+					<div class="form-group <%= 'form-group-error' if @errors.has_key? :person_name %> form-group--person_name">
+						<label class="form-label" for="person_name">
```
partial/helpers were created to render the form fields for the contact-us form.... these should probably be used on the signup form too for consistancy and to simplify the template

chrisfarms commented on assets/js/signup.js:
```diff
@@ -0,0 +1,176 @@
+$(function(){
+
+	var stepStack = [];
```
The back button is currently broken for the "js-enabled" version. This could be resolved (at least for modern browsers) by using the browser History API

chrisfarms replied:
wontfix yet

chrisfarms commented on app.rb:
```diff
@@ -0,0 +1,228 @@
+
+	set :sprockets, Sprockets::Environment.new
+	set :title, 'Gov.UK Platform as a Service'
+	enable :sessions
```
Sessions are no longer required ... hangover from the previous session-based forms

chrisfarms reviewed (changes requested):
Making some review notes myself of areas that could do with improvement.

Notes from review of forms with jess/stephen:

* Use quotes not italics
* Standardise on "org manager" whenever referencing
* Standardise on "organisation" (lowercase)
* we can drop the "name" from invites
* really want to be able to "add more" invites, even if only for the js-enabled version

46bit commented on Makefile:
```diff
@@ -0,0 +1,27 @@
+	@which -s npm || (echo 'npm is required to install the upstream sass dependencies. Install node js + npm.' && false)
+	npm install
+
+vendor: # reinstall ruby dependencies
```
`make vendor` doesn't run because a `vendor` directory already exists. This target needs to be [phony](https://www.gnu.org/software/make/manual/html_node/Phony-Targets.html).

chrisfarms replied:
You should run `make clean` if the target is stale. Making a target phony would make it run on every invocation.

46bit replied:
We've talked past each other a little. The reason this came up is that running `make vendor node_modules` on a fresh checkout says both are up to date. That happens because `vendor` and `node_modules` are checked in to the repo and thus already exist. Their contents are already ignored, so removing them from the repo (run `make clean` and commit then) would help?

46bit replied:
With @chrisfarms away today, I'll resolve this so we can get this into approving.

46bit replied:
Hmm, it was already no longer true that `vendor` and node_modules are checked in. Sorry Chris. Approving.

46bit commented on Makefile:
```diff
@@ -0,0 +1,27 @@
+help: ## Show this help.
+	@grep -E '^[a-zA-Z_-]+:.*?## .*$$' $(MAKEFILE_LIST) | awk 'BEGIN {FS = ":.*?## "}; {printf "\033[36m%-30s\033[0m %s\n", $$1, $$2}'
+
+node_modules: ## reinstall sass dependencies
```
`make node_modules` doesn't run because a `node_modules` directory already exists. This target needs to be [phony](https://www.gnu.org/software/make/manual/html_node/Phony-Targets.html).

46bit commented on Gemfile:
```diff
@@ -1,14 +1,29 @@
-# If you do not have OpenSSL installed, change
-# the following line to use 'http://'
+ruby '2.4.1'
```
It's common to also have a `.ruby-version` file that's used by Ruby version managers. In this case it'd contain `2.4.1\n` (if using `rbenv` you can also create it by typing `rbenv local 2.4.1`.)

I asked in `#ruby` on GDS Slack and it was suggested that, to avoid having the version in two different places, the `Gemfile` read the version from `.ruby-version`, like https://github.com/alphagov/smart-answers/blob/master/Gemfile#L3.

46bit commented on README.md:
```diff
@@ -2,55 +2,36 @@
 This is the source for the PaaS product page at https://www.cloud.service.gov.uk/
 
-It is based on: https://github.com/alphagov/product-page-example
+It is a Ruby/Sinatra app that makes use of the Gov.UK Elements, Frontend Toolkit and Template libraries.
```
It should be GOV.UK (all capitals.)

46bit commented on README.md:
```diff
@@ -2,55 +2,36 @@
-- `bower install`
-- `bundle exec middleman server`
-- `open http://localhost:4567`
+The following environment variables should be set for correct deployment
```
Needs a trailing `:`.

46bit commented on README.md:
```diff
@@ -2,55 +2,36 @@
+## Development
 
-To make use of it, follow these steps:
+Dependecies:
```
Dependencies

46bit commented on README.md:
```diff
@@ -2,55 +2,36 @@
+Dependecies:
 
-1. Build the docker image
+* Ruby
```
Ruby and Bundler

46bit commented on app.rb:
```diff
@@ -0,0 +1,228 @@
+class App < Sinatra::Base
+
+	set :sprockets, Sprockets::Environment.new
+	set :title, 'Gov.UK Platform as a Service'
```
GOV.UK (all capitalised.)

46bit commented on package.json:
```diff
@@ -0,0 +1,20 @@
+{
+  "name": "paas-product-page",
+  "version": "1.0.0",
+  "description": "product pages for Gov.UK PaaS",
```
GOV.UK (all capitalised.)

46bit commented on package.json:
```diff
@@ -0,0 +1,20 @@
+    "type": "git",
+    "url": "git+https://github.com/alphagov/paas-product-page.git"
+  },
+  "author": "Gov.UK",
```
GOV.UK (all capitalised.)

46bit reviewed (changes requested):
I'd prefer to re-review once my and @chrisfarms requested changes have been made.

46bit commented:
The DeskPro tickets that get submitted don't seem to differentiate between creating a new org and joining an existing one. Perhaps another boolean flag to put in the submitted JSON?

46bit commented:
On the `/signup` page, we're displaying the "Do you want to create an organisation?" section on the form when JavaScript is disabled. There's already *"If your team is already using GOV.UK PaaS and you would like a user account, ask your Org Manager to contact us and make a request at: gov-uk-paas-support@digital.cabinet-office.gov.uk"* text when JavaScript is disabled, so we'll want this section of the form to disappear.

Commit:
Port to Sinatra app and merge prototype changes

This is an unfortunely large commit which is the result of three
seperate large changes:

1. The product pages app was a static HTML app rendered with
   middleman and deployed with the static-buildpack (nginx),
   however we now want to app dynamic pages for form submissions.
   Sinatra was chosen for the backend since we did not need the
   full weight of Rails but wanted to use something other
   front-end developers at GDS would be comfortable with.
   * We simplified the stylesheet (sass) compiliation by
     keeping all required assets from GOV.UK elements/toolkit etc
     in the /assets dir and letting sprockets compile it JIT
   * Redirects were made to strip .html extensions for prettier URLs
   * DRYed up several repeated areas (like breadcrumbs) that were
     getting out of sync.
   * Added improved 404 pages
   * Added tests to catch dead links (since several were found).
   * Updated deployment scripts

2. Many long-lived changes from the "pass-proto" branch have been
   merged:
   * Many typos and html errors fixed
   * New home page design
   * New images
   * Content improvements based on user research
   * Updated support information

3. New "contact-us" and "signup" forms added:
   * Added new "get-started" onboarding page to guide
     potential new tennants.
   * Added a form for collecting the data required for new tennants.
   * Added progressive-enhancement of form when js-enabled so that:
        - The form is layed out as a "wizard" in multiple steps
        - There is clientside validation as you go
   * Added contact-us form for collecting support enquiries
   * Implemented a deskpro client library
   * Wired up the form submissions to send data to deskpro
   * Added mocked tests to ensure deskpro lib does it's job
   * Added tests to ensure form submissions are working

Commit:
Use a .ruby-version file

Using a .ruby-version file as the canonical source of which ruby version
to use (by reading this file from Gemfile too) make life easier for
people using rbenv and friends.

46bit commented:
We're going to iterate more on the product page over time, but this is definitely good for now. LGTM.
